### PR TITLE
[Assets] rpc assets, ipfs verification

### PR DESCRIPTION
**Pull request by nandofw**
added rpc cmd listunspentassets
added asset support in createrawtransaction
added ipfs verification
fixed asset test failure on windows

### CODE REVIEW

File: src/evo/providertx.cpp [hunk 1]
```diff
@@ -178,6 +178,13 @@ bool CheckNewAssetTx(const CTransaction &tx, const CBlockIndex *pindexPrev, CVal
         return state.DoS(100, false, REJECT_INVALID, "bad-assets-collateralAddress");
     }
 
+    if (assetTx.referenceHash.length() != 0){
+        if (assetTx.referenceHash.length() == 46){
+            if (assetTx.referenceHash.substr(0, 2) != "Qm")
+                return state.DoS(100, false, REJECT_INVALID, "bad-assets-referenceHash");
+        }
+        return state.DoS(100, false, REJECT_INVALID, "bad-assets-referenceHash");
+    }
```
Comment (rumaster) [inline src/evo/providertx.cpp]:
```suggestion
    if (assetTx.referenceHash.length() != 0)
        return state.DoS(100, false, REJECT_INVALID, "bad-assets-referenceHash");

    if (assetTx.referenceHash.length() == 46 && assetTx.referenceHash.substr(0, 2) != "Qm")
        return state.DoS(100, false, REJECT_INVALID, "bad-assets-referenceHash");
```

File: src/rpc/rawtransaction_util.cpp [hunk 2]
```diff
@@ -117,34 +119,71 @@ ConstructTransaction(const UniValue &inputs_in, const UniValue &outputs_in, cons
...
-                if (hasFuture) {
-                    throw JSONRPCError(RPC_INVALID_PARAMETER, std::string("can only send future to one address"));
+            bool hasasset = false;
+            std::string assetId;
+            uint32_t uniqueId;
+            if(sendToValue.isObject()) {
+                if (!sendToValue["assetid"].isNull()){
+                    hasasset = true;
+                    assetId = sendToValue["assetid"].get_str();
+                    if (!sendToValue["uniqueid"].isNull()){
```
Comment (rumaster) [inline src/rpc/rawtransaction_util.cpp:129]:
```suggestion
                if (!sendToValue["assetid"].isNull()) {
                    hasasset = true;
                    assetId = sendToValue["assetid"].get_str();
                    if (!sendToValue["uniqueid"].isNull()) {
```

File: src/rpc/rawtransaction_util.cpp [hunk 3]
```diff
@@ -117,34 +119,71 @@ ConstructTransaction(const UniValue &inputs_in, const UniValue &outputs_in, cons
...
+                        uniqueId = (uint32_t)sendToValue["uniqueid"].get_int64();
+                        nAmount = 1 * COIN;
+                    } else {
+                        if(sendToValue["amount"].isNull()) {
```
Comment (rumaster) [inline src/rpc/rawtransaction_util.cpp]:
```suggestion
            if (sendToValue.isObject()) {
                if (!sendToValue["assetid"].isNull()) {
                    hasasset = true;
                    assetId = sendToValue["assetid"].get_str();
                    if (!sendToValue["uniqueid"].isNull()) {
                        uniqueId = (uint32_t)sendToValue["uniqueid"].get_int64();
                        nAmount = 1 * COIN;
                    } else {
                        if (sendToValue["amount"].isNull()) {
```

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -417,16 +417,17 @@ UniValue sendtoaddress(const JSONRPCRequest &request) {
...
             fpp.locktime = request.params[nextParamsIndex]["future_locktime"].get_int();
-            fpp.futureRecScript = GetScriptForDestination(dest);
-            nextParamsIndex++;
+            fpp.futureRecScript = GetScriptForDestination(dest);           
         }
     }
+    nextParamsIndex++;
     if (request.params[nextParamsIndex].isStr() && !request.params[nextParamsIndex].get_str().empty()) {
-        mapValue["comment"] = request.params[nextParamsIndex].get_str();
+        mapValue["comment"] = request.params[nextParamsIndex].get_str();
```
Comment (rumaster) [inline src/wallet/rpcwallet.cpp]:
```suggestion
        mapValue["comment"] = request.params[nextParamsIndex].get_str();
```

File: src/rpc/rpcassets.cpp [hunk 5]
```diff
@@ -610,6 +810,7 @@ static const CRPCCommand commands[] =
                 {"assets", "sendasset",         &sendasset,         {"assetId", "amount", "address", "change_address", "asset_change_address"}},
                 {"assets", "assetdetails",      &assetdetails,      {"assetId"}},
                 {"assets", "listassetsbalance", &listassetsbalance, {}},
+                {"assets", "listunspentassets", &listunspentassets, {"minconf","maxconf","addresses","include_unsafe","query_options"}},
```
Comment (rumaster) [inline src/rpc/rpcassets.cpp]:
```suggestion
                {"assets", "listunspentassets", &listunspentassets, {"minconf", "maxconf", "addresses", "include_unsafe", "query_options"}},
```

File: src/evo/providertx.cpp (hunk 1 above)
Comment (nandofw) [reply src/evo/providertx.cpp]: this change will reject any referenceHash that has length greater than 0.
valid referenceHash can have length 0 or length 46 and start with "Qm"

File: src/test/assets_tests.cpp [hunk 6]
```diff
@@ -81,6 +81,7 @@ static void FundTransaction(CMutableTransaction &tx, SimpleUTXOMap &utoxs, const
 }
 
 static void SignTransaction(const CTxMemPool &mempool, CMutableTransaction &tx, const CKey &coinbaseKey) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(50)); //txindex extra time for windows
```
Comment (rumaster) [inline src/test/assets_tests.cpp:84]: looks like a crutch

**Review by npq7721 — APPROVED** [review]
LGTM

File: src/evo/providertx.cpp (hunk 1 above)
Comment (rumaster) [reply src/evo/providertx.cpp]: This is the equivalent. At line 189, any referenceHash that has a length greater than 0 will be rejected.

Comment (rumaster) [reply src/evo/providertx.cpp]: yeah... I hesitated for a moment..
This is the absolute equivalent. At line 189, any referenceHash that has a length greater than 0 will be rejected.

Comment (rumaster) [reply src/evo/providertx.cpp]: it looks like it passes with length 0 only. ((

**Pull request merged**